presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeFilter.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.presto.metadata.FunctionRegistry;
+import com.facebook.presto.sql.tree.DefaultExpressionTraversalVisitor;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.GroupingOperation;
+import com.facebook.presto.sql.tree.Node;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public final class ExpressionTreeFilter
+{
+    private ExpressionTreeFilter() {}
+
+    public static Function<Node, Optional<Expression>> groupingExtractionFunction = (node -> {
+        if (node instanceof GroupingOperation) {
+            return Optional.of((GroupingOperation) node);
+        }
+        return Optional.empty();
+    });
+
+    public static Function<Node, Optional<Expression>> windowExtractorFunction = (node -> {
+        if (node instanceof FunctionCall) {
+            FunctionCall functionCall = (FunctionCall) node;
+            if (functionCall.getWindow().isPresent()) {
+                return Optional.of(functionCall);
+            }
+        }
+        return Optional.empty();
+    });
+
+    public static Function<Node, Optional<Expression>> aggregateExtractorFunction(FunctionRegistry functionRegistry)
+    {
+        return ((node) -> {
+            if (node instanceof FunctionCall) {
+                FunctionCall functionCall = (FunctionCall) node;
+                if ((functionRegistry.isAggregationFunction(functionCall.getName()) || functionCall.getFilter().isPresent()) && !functionCall.getWindow().isPresent()) {
+                    return Optional.of(functionCall);
+                }
+            }
+            return Optional.empty();
+        });
+    }
+
+    public static List<Expression> extract(Function<Node, Optional<Expression>> predicate, List<? extends Node> nodes)
```

Comment:
1. The advantage of `Function` is that you can specify a type for both the input and output so it's probably a little safer than `Predicate`.
2. Unfortunately, some of the inputs that need to be analyzed derive from `Node` and not from `Expression`. For example, see `StatementAnalyzer#hasAggregates()`, which feeds `SingleColumn` and `SortItem` objects to the extractor. I could extract the expression portions of those objects but I'd rather that be done by the `linearizeNodes()` function.
3. I've refactored the method as we agreed offline.